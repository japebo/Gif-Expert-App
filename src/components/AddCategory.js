import React, { useState } from 'react'

export const AddCategory = ( { setCategories } ) => {
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = ( e ) => {
        // console.log(e.target.value);
        setInputValue(e.target.value); //e.target.value contiene el valor tras ocurrir el evento de presion de una tecla.
    }

    const handleSubmit = (e) => {
        e.preventDefault(); //evita que se recargue la página, que es el comportamiento por defecto tras el evento.

        if (inputValue.trim().length > 2) {
            setCategories( cats => [...cats, inputValue]); //opción preferida, que es usando un callback function, pues setCategories tiene acceso al esatdo actual de las 'categories', accesible en este caso a través del argumento 'cats' que pasa la propia función.
            // setCategories( [...categories, inputValue]); //una alternativa a la línea de arriba es esta, que hace lo mismo. Para usarla habría que pasarle la propiedad 'categories' (o el nombre que se desee, solo debemos asegurarnos de que coincidan el nombre de las propiedad pasada y el argumento que se desestructura luego en el functional component AddCategory) al llamado al componente AddCategory dentro del componente GifExpertApp (de esta forma: <AddCategory setCategories={setCategories} categories={categories}/>); además de desestructurar la nueva propiedad que es pasada de un componente a otro, llamada categories probablemente (de esta forma: export const AddCategory = ( { categories, setCategories  } ) => {...}). 
            setInputValue('');
        };
        // console.log('Submit hecho');
    };

    return (
        <form onSubmit = {handleSubmit}>
           <input
                type="text"
                value={ inputValue }
                onChange = {handleInputChange}
           /> 
        </form>
    )
}
