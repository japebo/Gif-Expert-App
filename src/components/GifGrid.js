
import React, {useState, useEffect} from 'react';
import { GifGridItem } from './GifGridItem';
import { getGifs } from '../helpers/getGifs';

export const GifGrid = ({category}) => {

    const [images, setImages] = useState([]);
 
    useEffect(() => { //hook
        getGifs(category).then( imgs => setImages(imgs) ); //helper that returns a promise
    }, [category]); //saying that if category changes (that will not), the instrucción inside useEffect will be re-runned.

    return (
        <>
            <h3> {category} </h3>
            <div className="card-grid">
                {
                    images.map( img => (
                        // <GifGridItem key={img.id} img={img} />  //opción 1
                        <GifGridItem key={img.id} {...img} />   //opción 2 
                        ))  
                    }
            </div>
        </>
    )
}
