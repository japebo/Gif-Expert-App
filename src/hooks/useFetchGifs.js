import { useEffect, useState } from "react";
import { getGifs } from '../helpers/getGifs'

export const useFetchGifs = (category) => { //this is a custom hook

    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect(() => { 
        getGifs(category).then( imgs => setState({
                data: imgs,
                loading: false    
            }) ); //helper that returns a promise
    }, [category]); //saying that if category changes (that will not), the instrucción inside useEffect will be re-runned.
    // en el hook de arriba, la función del primer argumento no puede ser async, porque espera algo síncrono    

    return state;
}