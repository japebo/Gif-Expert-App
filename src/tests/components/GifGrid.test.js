import React from 'react';
import { shallow } from "enzyme"
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs'); //crea un mock del custom hook (función) para mimificarla y por ejemplo fingir que retornó valores que se definen aquí mismo en la sección de pruebas

describe('Pruebas en <GifGrid />', () => {
    const category = 'Sponge Bob';

    test('Debe hacer match con el Snapshot', () => {
        useFetchGifs.mockReturnValue({ 
            data: [],
            loading: true
        });
        const wrapper = shallow(<GifGrid category={category}/>);
        expect(wrapper).toMatchSnapshot();        
    })
    
    test('Debe mostrar items cuando se cargan imágenes useFetchGifs', () => {

        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/cualquier/cosa.jpg',
            title: 'Cualquier cosa' 
        }];

        useFetchGifs.mockReturnValue({ //fingiendo que el custom hook useFetchGifs retornó estos valores para poder tener un snapshot con estos valores
            data: gifs,
            loading: false
        });
        const wrapper = shallow(<GifGrid category={category}/>); //renderización del componente
        
        expect(wrapper).toMatchSnapshot();
        expect( wrapper.find('p').exists() ).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length); //para comprobar que la cantidad de componentes que se están renderizando sea igual a la cantidad de objetos que retorna el helper getGifs() en el array
    })
})
