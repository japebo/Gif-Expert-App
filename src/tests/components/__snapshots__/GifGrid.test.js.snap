// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid /> Debe hacer match con el Snapshot 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
     
    Sponge Bob
     
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Pruebas en <GifGrid /> Debe mostrar items cuando se cargan im√°genes useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
     
    Sponge Bob
     
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualquier cosa"
      url="https://localhost/cualquier/cosa.jpg"
    />
  </div>
</Fragment>
`;
